File: Task-Manager/taskmanager-backend/taskmanager-backend/src/main/java/com/taskmanager/service/TaskService.java
package com.taskmanager.service;

import com.taskmanager.model.Task;
import com.taskmanager.model.Task.TaskStatus;
import com.taskmanager.model.Task.TaskPriority;
import com.taskmanager.model.User;
import com.taskmanager.repository.TaskRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

@Service
@Transactional

public class TaskService {
    
    @Autowired
    private TaskRepository taskRepository;
    
    // Get all tasks for a user
    public List<Task> getAllTasks(User user) {
        return taskRepository.findByUserAndDeletedAtIsNullOrderByCreatedAtDesc(user);
    }
    
    // Get all tasks with pagination
    public Page<Task> getAllTasksPaginated(User user, Pageable pageable) {
        return taskRepository.findByUserAndDeletedAtIsNull(user, pageable);
    }
    
    // Get task by ID
    public Optional<Task> getTaskById(Long id, User user) {
        return taskRepository.findByIdAndUser(id, user);
    }
    
    // Get tasks by status
    public List<Task> getTasksByStatus(User user, TaskStatus status) {
        return taskRepository.findByUserAndStatusAndDeletedAtIsNull(user, status);
    }
    
    // Get tasks by priority
    public List<Task> getTasksByPriority(User user, TaskPriority priority) {
        return taskRepository.findByUserAndPriorityAndDeletedAtIsNull(user, priority);
    }
    
    // Search tasks
    public List<Task> searchTasks(User user, String keyword) {
        return taskRepository.findByUserAndTitleContainingIgnoreCaseAndDeletedAtIsNull(user, keyword);
    }
    
    // Advanced filtering
    public Page<Task> filterTasks(User user, TaskStatus status, TaskPriority priority, 
                                  Boolean archived, String keyword, Pageable pageable) {
        return taskRepository.findByFilters(user, status, priority, archived, keyword, pageable);
    }
    
    // Create task
    public Task createTask(Task task, User user) {
        if (task.getStatus() == null) {
            task.setStatus(TaskStatus.TODO);
        }
        if (task.getPriority() == null) {
            task.setPriority(TaskPriority.MEDIUM);
        }
        task.setUser(user);
        return taskRepository.save(task);
    }
    
    // Update task
    public Task updateTask(Long id, Task taskDetails, User user) {
        Task task = taskRepository.findByIdAndUser(id, user)
            .orElseThrow(() -> new RuntimeException("Task not found with id: " + id));
        
        task.setTitle(taskDetails.getTitle());
        task.setDescription(taskDetails.getDescription());
        task.setStatus(taskDetails.getStatus());
        task.setPriority(taskDetails.getPriority());
        task.setDueDate(taskDetails.getDueDate());
        
        return taskRepository.save(task);
    }
    
    // Update task status
    public Task updateTaskStatus(Long id, TaskStatus status, User user) {
        Task task = taskRepository.findByIdAndUser(id, user)
            .orElseThrow(() -> new RuntimeException("Task not found with id: " + id));
        task.setStatus(status);
        return taskRepository.save(task);
    }
    
    // Archive task
    public Task archiveTask(Long id, User user) {
        Task task = taskRepository.findByIdAndUser(id, user)
            .orElseThrow(() -> new RuntimeException("Task not found with id: " + id));
        task.setArchived(true);
        return taskRepository.save(task);
    }
    
    // Unarchive task
    public Task unarchiveTask(Long id, User user) {
        Task task = taskRepository.findByIdAndUser(id, user)
            .orElseThrow(() -> new RuntimeException("Task not found with id: " + id));
        task.setArchived(false);
        return taskRepository.save(task);
    }
    
    // Get archived tasks
    public List<Task> getArchivedTasks(User user) {
        return taskRepository.findByUserAndArchivedAndDeletedAtIsNull(user, true);
    }
    
    // Soft delete
    public void deleteTask(Long id, User user) {
        Task task = taskRepository.findByIdAndUser(id, user)
            .orElseThrow(() -> new RuntimeException("Task not found with id: " + id));
        task.setDeletedAt(LocalDateTime.now());
        taskRepository.save(task);
    }
    
    // Get deleted tasks
    public List<Task> getDeletedTasks(User user) {
        return taskRepository.findByUserAndDeletedAtIsNotNull(user);
    }
    
    // Restore task
    public Task restoreTask(Long id, User user) {
        Task task = taskRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Task not found with id: " + id));
        
        if (!task.getUser().getId().equals(user.getId())) {
            throw new RuntimeException("Unauthorized");
        }
        
        if (task.getDeletedAt() == null) {
            throw new RuntimeException("Task is not deleted");
        }
        
        task.setDeletedAt(null);
        return taskRepository.save(task);
    }
    
    // Permanent delete
    public void permanentlyDeleteTask(Long id, User user) {
        Task task = taskRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Task not found with id: " + id));
        
        if (!task.getUser().getId().equals(user.getId())) {
            throw new RuntimeException("Unauthorized");
        }
        
        taskRepository.delete(task);
    }
    
    // Delete all tasks
    public void deleteAllTasks(User user) {
        List<Task> tasks = taskRepository.findByUserAndDeletedAtIsNull(user);
        LocalDateTime now = LocalDateTime.now();
        tasks.forEach(task -> task.setDeletedAt(now));
        taskRepository.saveAll(tasks);
    }
}

File: Task-Manager/taskmanager-backend/taskmanager-frontend/src/app/services/task.service.ts
// src/app/services/task.service.ts

import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { PagedTasks, Task, TaskPriority, TaskStatus } from '../models/task.model';

@Injectable({
  providedIn: 'root'
})
export class TaskService {
  private apiUrl = 'http://localhost:8080/api/tasks';

  constructor(private http: HttpClient) { }

  // Get all tasks
  getAllTasks(): Observable<Task[]> {
    return this.http.get<Task[]>(this.apiUrl);
  }

  // Get task by ID
  getTaskById(id: number): Observable<Task> {
    return this.http.get<Task>(`${this.apiUrl}/${id}`);
  }

  // Get tasks by status
  getTasksByStatus(status: TaskStatus): Observable<Task[]> {
    return this.http.get<Task[]>(`${this.apiUrl}/status/${status}`);
  }

  // Search tasks by keyword
  searchTasks(keyword: string): Observable<Task[]> {
    return this.http.get<Task[]>(`${this.apiUrl}/search?keyword=${keyword}`);
  }

  // Create new task
  createTask(task: Task): Observable<Task> {
    return this.http.post<Task>(this.apiUrl, task);
  }

  // Update existing task
  updateTask(id: number, task: Task): Observable<Task> {
    return this.http.put<Task>(`${this.apiUrl}/${id}`, task);
  }

  // Update task status only
  updateTaskStatus(id: number, status: TaskStatus): Observable<Task> {
    return this.http.patch<Task>(`${this.apiUrl}/${id}/status`, { status });
  }

  // Delete task
  deleteTask(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  // ===== NEW METHODS FOR PHASE 1 =====

  // Get tasks with pagination
  getTasksPaginated(page: number = 0, size: number = 10, sortBy: string = 'createdAt', sortDir: string = 'desc'): Observable<PagedTasks> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sortBy', sortBy)
      .set('sortDir', sortDir)
      .set('paginate', 'true');
    
    return this.http.get<PagedTasks>(this.apiUrl, { params });
  }

  // Advanced filtering
  filterTasks(
    status?: TaskStatus,
    priority?: TaskPriority,
    archived?: boolean,
    keyword?: string,
    page: number = 0,
    size: number = 10
  ): Observable<PagedTasks> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());

    if (status) params = params.set('status', status);
    if (priority) params = params.set('priority', priority);
    if (archived !== undefined) params = params.set('archived', archived.toString());
    if (keyword) params = params.set('keyword', keyword);

    return this.http.get<PagedTasks>(`${this.apiUrl}/filter`, { params });
  }

  // Get archived tasks
  getArchivedTasks(): Observable<Task[]> {
    return this.http.get<Task[]>(`${this.apiUrl}/archived`);
  }

  // Get deleted tasks
  getDeletedTasks(): Observable<Task[]> {
    return this.http.get<Task[]>(`${this.apiUrl}/deleted`);
  }

  // Archive task
  archiveTask(id: number): Observable<Task> {
    return this.http.patch<Task>(`${this.apiUrl}/${id}/archive`, {});
  }

  // Unarchive task
  unarchiveTask(id: number): Observable<Task> {
    return this.http.patch<Task>(`${this.apiUrl}/${id}/unarchive`, {});
  }

  // Restore deleted task
  restoreTask(id: number): Observable<Task> {
    return this.http.patch<Task>(`${this.apiUrl}/${id}/restore`, {});
  }

  // Permanent delete
  permanentlyDeleteTask(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}/permanent`);
  }

  // Get tasks by priority
  getTasksByPriority(priority: TaskPriority): Observable<Task[]> {
    return this.http.get<Task[]>(`${this.apiUrl}/priority/${priority}`);
  }
}